app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt

```diff
@@ -20,6 +30,76 @@ class Preferences : PreferenceFragmentCompat() {
             it.summary = path
         }
     }
+
+    override fun onPreferenceTreeClick(preference: Preference): Boolean {
+        val sharedPreferences = context?.let { PreferenceManager.getDefaultSharedPreferences(it) }
+
+        if (preference.key == "enable_dnd") { // If someone toggled the enable_dnd preference
+            if (sharedPreferences != null) {
+                if (sharedPreferences.getBoolean("enable_dnd", false)) {
+                    val activity = activity ?: return super.onPreferenceTreeClick(preference)
+                    val noteServ = Context.NOTIFICATION_SERVICE // 100 char limit
+                    val dndManager = activity.getSystemService(noteServ) as NotificationManager
+                    val hasPermission = dndManager.isNotificationPolicyAccessGranted
+
+                    if (!hasPermission) {
+                        // If we don't have permissions for DND
+                        AlertDialog.Builder(context)
+                            .setTitle(R.string.settings_enable_dnd_q_title)
+                            .setMessage(R.string.settings_enable_dnd_q_message)
+                            .setPositiveButton(R.string.settings_enable_dnd_q_ok) { _, _ ->
+                                val setting_id = Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS
+                                val intent = Intent(setting_id)
+                                startActivity(intent)
+                            }
+                            .setNegativeButton(R.string.settings_enable_dnd_q_cancel) { _, _ ->
+                                val editor = sharedPreferences.edit()
+                                if (editor != null) {
+                                    editor.putBoolean("enable_dnd", false)
+                                    editor.apply()
+                                } else {
+                                    Log.wtf(TAG, "editor is null")
+                                }
+                                checkDnd()
+                            }
+                            .create()
+                            .show()
+                    }
+                }
+            } else {
+                Log.wtf(TAG, "sharedPreferences is null")
+            }
+        }
+        return super.onPreferenceTreeClick(preference)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        checkDnd()
+    }
+
+    fun checkDnd() {
+        val activity = activity ?: return
+        val noteServ = Context.NOTIFICATION_SERVICE // 100 char limit
+        val dndManager = activity.getSystemService(noteServ) as NotificationManager
+        val hasPermission = dndManager.isNotificationPolicyAccessGranted
+
+        val sharedPreferences = context?.let { PreferenceManager.getDefaultSharedPreferences(it) }
+        val editor = sharedPreferences?.edit()
+
+        if (!hasPermission) {
+            if (editor != null) {
+                editor.putBoolean("enable_dnd", false)
```

Comment:
It would be also an option to just not to do this. The "enabled, but no permission" case is something you have to handle anyway (and you do), so if the user explicitly taps enable + rejects the permission, I would not auto-disable.